### PR TITLE
Rework ADMatDiffusion and add anisotropic version

### PR DESCRIPTION
**Pull request by dschwen**
Let's see what this breaks... (ping @tophmatthews )

Refs #13632

### CODE REVIEW

**Comment by dschwen** [comment]
The Bison failures are simple fixes (parameter name change). The other failures seem to be related to @permcody's Makefile changes (now mixed double and single `:` rules).

**Comment by dschwen** [comment]
I do like `diffusivity` better than `D_name` for that parameter. I'm thinking we should rather rename the parameter in `MatDiffusion` than change it here. (That way I wouldn't have to update bison either.... but probably a few other apps :-) )

**Comment by tophmatthews** [comment]
That “D_name” drives me crazy.

File: modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  [./D]
+    type = ConstantAnisotropicMobility
+    tensor = '0.1 0 0
+              0   1 0
+              0   0 0'
+    M_name = D
+  [../]
+[]
+
+[Preconditioning]
```
Comment (tophmatthews) [inline modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i]: Not necessary to provide preconditioning block when using newton now

File: modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   full = true
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  scheme = 'BDF2'
+
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -pc_asm_overlap'
```
Comment (tophmatthews) [inline modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i]: are all these executioner pieces necessary for the solve?

File: modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  ny = 15
+  xmax = 15.0
+  ymax = 15.0
+  elem_type = QUAD4
+[]
+
+[Variables]
+  [./c]
+    [./InitialCondition]
+      type = CrossIC
```
Comment (tophmatthews) [inline modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i]: Nice test for non-linearity

**Review by tophmatthews — COMMENTED** [review]
Just small things

File: framework/include/kernels/ADMatDiffusionBase.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  const ADVariableGradient & _grad_conc;
+
+  usingKernelGradMembers;
+};
+
+template <ComputeStage compute_stage, typename T>
+ADMatDiffusionBase<compute_stage, T>::ADMatDiffusionBase(const InputParameters & parameters)
+  : ADKernelGrad<compute_stage>(parameters),
+    _D(adGetADMaterialProperty<T>("D_name")),
+    _grad_conc(isCoupled("conc") ? adCoupledGradient("conc") : _grad_u)
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/include/kernels/ADMatDiffusionBase.h]: We use `v` a lot in those cases

**Comment by dschwen** [comment]
Some of the app test failures are related to the consistent memberfunction access PR. The bison and marmot fixes should be quick.

File: framework/doc/content/source/kernels/ADMatDiffusion.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# ADMatDiffusion
+
+## Description
+
+`ADMatDiffusion` implements the term
+\begin{equation}
+\nabla D(a,b,\dots) \nabla c,
```
Comment (tophmatthews) [inline framework/doc/content/source/kernels/ADMatDiffusion.md]: \nabla u?

File: framework/include/kernels/ADMatDiffusionBase.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADKernelGrad.h"
+
+#define usingMatDiffusionBaseMembers(T)                                                            \
+  usingKernelGradMembers;                                                                          \
+  using ADMatDiffusionBase<compute_stage, T>::_diffusivity;                                        \
```
Comment (tophmatthews) [inline framework/include/kernels/ADMatDiffusionBase.h]: 👍

File: framework/include/kernels/MatDiffusionBase.h [hunk 7]
```diff
@@ -52,20 +52,27 @@ class MatDiffusionBase : public DerivativeMaterialInterface<JvarMapKernelInterfa
...
 };
 
 template <typename T>
 InputParameters
 MatDiffusionBase<T>::validParams()
 {
   InputParameters params = ::validParams<Kernel>();
-  params.addParam<MaterialPropertyName>("D_name", "D", "The name of the diffusivity");
+  params.addDeprecatedParam<MaterialPropertyName>(
+      "D_name",
```
Comment (tophmatthews) [inline framework/include/kernels/MatDiffusionBase.h]: 👍

File: framework/include/kernels/MatDiffusionBase.h [hunk 8]
```diff
@@ -52,20 +52,27 @@ class MatDiffusionBase : public DerivativeMaterialInterface<JvarMapKernelInterfa
...
   /// Gradient of the concentration
-  const VariableGradient & _grad_conc;
+  const VariableGradient & _grad_v;
 };
 
 template <typename T>
 InputParameters
 MatDiffusionBase<T>::validParams()
 {
   InputParameters params = ::validParams<Kernel>();
```
Comment (tophmatthews) [inline framework/include/kernels/MatDiffusionBase.h]: What do the double colons do here?

File: framework/include/kernels/MatDiffusionBase.h [hunk 9]
```diff
@@ -52,20 +52,27 @@ class MatDiffusionBase : public DerivativeMaterialInterface<JvarMapKernelInterfa
...
-  params.addParam<MaterialPropertyName>("D_name", "D", "The name of the diffusivity");
+  params.addDeprecatedParam<MaterialPropertyName>(
+      "D_name",
+      "The name of the diffusivity",
+      "This parameter has been renamed to 'diffusivity', which is more mnemonic and more conducive "
+      "to passing a number literal");
+  params.addParam<MaterialPropertyName>(
+      "diffusivity", "D", "The diffusivity value or material property");
```
Comment (tophmatthews) [inline framework/include/kernels/MatDiffusionBase.h]: Is this the best default? I would think "diffusivity" would be better, same as the name.

File: framework/include/kernels/MatDiffusionBase.h [hunk 10]
```diff
@@ -52,20 +52,27 @@ class MatDiffusionBase : public DerivativeMaterialInterface<JvarMapKernelInterfa
...
   params.addCoupledVar("args", "Vector of arguments of the diffusivity");
```
Comment (tophmatthews) [inline framework/include/kernels/MatDiffusionBase.h]: maybe, "Optional vector of arguments for the diffusivity. If provided and diffusivity is a derivative parsed material, Jacobian contributions from the diffusivity will be automatically computed"

Comment (tophmatthews) [reply framework/include/kernels/MatDiffusionBase.h]: Well, actually this pr isn't really about non-AD...just a suggestion

File: framework/src/kernels/ADMatDiffusionBase.C [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+defineADValidParams(
+    ADMatDiffusionBase,
+    ADKernelGrad,
+    params.addClassDescription("Diffusion kernel with a material property as diffusivity and "
+                               "automatic differentiation to provide perfect Jacobians");
+    params.addParam<MaterialPropertyName>("diffusivity",
+                                          "D",
+                                          "The diffusivity value or material property");
+    params.addCoupledVar("args", "Vector of arguments of the diffusivity");
```
Comment (tophmatthews) [inline framework/src/kernels/ADMatDiffusionBase.C]: Maybe use suggestion above on a more clear description?

File: modules/phase_field/test/tests/Nucleation/marker.i [hunk 12]
```diff
@@ -41,7 +41,7 @@
 [Kernels]
   [./c]
     type = Diffusion
-    #D_name = F
+    #diffusivity = F
```
Comment (tophmatthews) [inline modules/phase_field/test/tests/Nucleation/marker.i]: Probably can just drop the commented out diffusivity? Several more that follow too.

File: modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i [hunk 13]
```diff
@@ -0,0 +1,63 @@
+[Mesh]
+  type = GeneratedMesh
+  dim = 2
+  nx = 15
+  ny = 15
+  xmax = 15.0
+  ymax = 15.0
+  elem_type = QUAD4
```
Comment (tophmatthews) [inline modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i]: unnecessary elem_type?

**Review by tophmatthews — DISMISSED** [review]
Good idea, I like how you set these up to apply to isotropic and anisotropic easily

File: modules/phase_field/test/tests/Nucleation/marker.i (hunk 12 above)
Comment (dschwen) [reply modules/phase_field/test/tests/Nucleation/marker.i]: Ugh, if I change the default I'll probably have a ton of input files to fix...

File: modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i (hunk 13 above)
Comment (dschwen) [reply modules/phase_field/test/tests/anisotropic_mobility/ad_diffusion.i]: yeah, here and in hundreds of inputs. I'll remove this one

File: framework/include/kernels/MatDiffusionBase.h (hunk 10 above)
Comment (dschwen) [reply framework/include/kernels/MatDiffusionBase.h]: I like that

File: framework/include/kernels/MatDiffusionBase.h (hunk 8 above)
Comment (dschwen) [reply framework/include/kernels/MatDiffusionBase.h]: The resolve to the global namespace. This is needed to differentiate from the class's own validParams member function.

File: framework/doc/content/source/kernels/ADMatDiffusion.md (hunk 5 above)
Comment (dschwen) [reply framework/doc/content/source/kernels/ADMatDiffusion.md]: k

File: framework/src/kernels/ADMatDiffusionBase.C (hunk 11 above)
Comment (dschwen) [reply framework/src/kernels/ADMatDiffusionBase.C]: ha, `args` is not even needed in the AD version. I'll remove it.

**Comment by dschwen** [comment]
Ok, made the changes and improved the docs a bit (more verbose, and I had forgotten to update the parameter names...). I'm sitting on pins and needles here because I'd like to wrap up the Marmot PR that depends on this. No pressure though (just kidding, the pressure is on, totally!).

**Comment by dschwen** [comment]
Hold on, I screwed up my commit history a bit

File: framework/doc/content/source/kernels/ADMatDiffusion.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# ADMatDiffusion
+
+## Description
+
+`ADMatDiffusion` implements the term
+\begin{equation}
+\nabla D(a,b,\dots) \nabla u,
```
Comment (laagesen) [inline framework/doc/content/source/kernels/ADMatDiffusion.md]: please put \cdot between \nabla and D

File: framework/doc/content/source/kernels/MatDiffusion.md [hunk 15]
```diff
@@ -4,12 +4,18 @@
 
 Implements the term
 \begin{equation}
-\nabla D(c,a,b,\dots) \nabla c,
+\nabla D(c,a,b,\dots) \nabla u,
```
Comment (laagesen) [inline framework/doc/content/source/kernels/MatDiffusion.md]: please put \cdot between \nabla and D

**Review by laagesen — CHANGES_REQUESTED** [review]

**Review by laagesen — DISMISSED** [review]

**Review by tophmatthews — DISMISSED** [review]

**Comment by permcody** [comment]
@dschwen - We had a lot of updates yesterday to get the API incompatibility through the system and all apps. Can you rebase this on current devel so we can see another round of testing?

**Comment by dschwen** [comment]
Done. It's a little annoying that a simple rebasing invalidates reviews...

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
> Done. It's a little annoying that a simple rebasing invalidates reviews...

Yeah, they are new commits from Git's point-of-view so it's kinda hard to know when to invalidate versus not. Oh well, I saw the previous approvals and this looks fine to me. Merge when testing is complete.

**Comment by dschwen** [comment]
Thanks, I have Bison and Marmot fixes almost ready...

**Comment by permcody** [comment]
@dschwen - Your Marmot patch failed - looks like ambiguity in your Doc links? Let me know if you need help fixing.

**Comment by dschwen** [comment]
I'll see if I can fix it myself. Just walking back from hit rocks Cafe.

**Pull request merged**